### PR TITLE
fix: distinct doesn't work as a column wrapper

### LINKED ISSUE #188: `distinct` doesn't work as a column wrapper - no label is given
#### Environment details

  - OS type and version: macOs 11.4
  - Python version: Python 3.9
  - `pybigquery` 0.9.0 (also tested on 0.8.0)

#### Steps to reproduce

  make a query that uses `sqlalchemy.distinct` and not Query.distinct
 EDIT: found out that the problem is that the former does not add an `AS` to the query, hence - adding a label manually is a workaround
  
#### Code example

```python
  # this raises an error
  from sqlalchemy import distinct
  db.query(distinct(MyTable.my_column)).all()
  
  # this plays fine:
  db.query(MyTable.my_column).distinct().all()
  
  #edit: this ALSO plays nice
  from sqlalchemy import distinct
  db.query(distinct(MyTable.my_column).label('just_a_random_label').all()
```

#### Stack trace
```
Error on request:
Traceback (most recent call last):
  File "/Users/some/path/.venv/lib/python3.9/site-packages/sqlalchemy/engine/cursor.py", line 639, in _index_for_key
    rec = self._keymap[key]
KeyError: <sqlalchemy.sql.elements.UnaryExpression object at 0x12e0b9760>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/some/path/.venv/lib/python3.9/site-packages/werkzeug/serving.py", line 319, in run_wsgi
    execute(self.server.app)
  File "/Users/some/path/.venv/lib/python3.9/site-packages/werkzeug/serving.py", line 308, in execute
    application_iter = app(environ, start_response)
  File "/Users/some/path/.venv/lib/python3.9/site-packages/flask/app.py", line 2088, in __call__
    return self.wsgi_app(environ, start_response)
  File "/Users/some/path/.venv/lib/python3.9/site-packages/flask/app.py", line 2073, in wsgi_app
    response = self.handle_exception(e)
  File "/Users/some/path/.venv/lib/python3.9/site-packages/flask/app.py", line 2070, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/some/path/.venv/lib/python3.9/site-packages/flask/app.py", line 1515, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/some/path/.venv/lib/python3.9/site-packages/flask/app.py", line 1513, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/some/path/.venv/lib/python3.9/site-packages/flask/app.py", line 1499, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/Users/some/path/app/my_file.py", line 666, in my_function()
    industries = get_db_session().query(distinct(MyTable.column)).all()
  File "/Users/some/path/.venv/lib/python3.9/site-packages/sqlalchemy/orm/query.py", line 2699, in all
    return self._iter().all()
  File "/Users/some/path/.venv/lib/python3.9/site-packages/sqlalchemy/orm/query.py", line 2834, in _iter
    result = self.session.execute(
  File "/Users/some/path/.venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 1680, in execute
    result = compile_state_cls.orm_setup_cursor_result(
  File "/Users/some/path/.venv/lib/python3.9/site-packages/sqlalchemy/orm/context.py", line 318, in orm_setup_cursor_result
    return loading.instances(result, querycontext)
  File "/Users/some/path/.venv/lib/python3.9/site-packages/sqlalchemy/orm/loading.py", line 88, in instances
    cursor.close()
  File "/Users/some/path/.venv/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/Users/some/path/.venv/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/Users/some/path/.venv/lib/python3.9/site-packages/sqlalchemy/orm/loading.py", line 68, in instances
    *[
  File "/Users/some/path/.venv/lib/python3.9/site-packages/sqlalchemy/orm/loading.py", line 69, in <listcomp>
    query_entity.row_processor(context, cursor)
  File "/Users/some/path/.venv/lib/python3.9/site-packages/sqlalchemy/orm/context.py", line 2597, in row_processor
    getter = result._getter(column)
  File "/Users/some/path/.venv/lib/python3.9/site-packages/sqlalchemy/engine/result.py", line 881, in _getter
    return self._metadata._getter(key, raiseerr)
  File "/Users/some/path/.venv/lib/python3.9/site-packages/sqlalchemy/engine/result.py", line 98, in _getter
    index = self._index_for_key(key, raiseerr)
  File "/Users/some/path/.venv/lib/python3.9/site-packages/sqlalchemy/engine/cursor.py", line 641, in _index_for_key
    rec = self._key_fallback(key, ke, raiseerr)
  File "/Users/some/path/.venv/lib/python3.9/site-packages/sqlalchemy/engine/cursor.py", line 783, in _key_fallback
    util.raise_(
  File "/Users/some/path/.venv/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
sqlalchemy.exc.NoSuchColumnError: "Could not locate column in row for column 'DISTINCT my_table.my_column'"
```

Making sure to follow these steps will guarantee the quickest resolution possible.

Thanks!

### PR DESCRIPTION
**Pull request by jimfulton**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-sqlalchemy/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #188  🦕

### CODE REVIEW

**Comment by jimfulton** [comment]
Note to reviewers:

This bug occurred because the dialect copies `visit_bindparam` from SQLAlchemy. See: https://github.com/googleapis/python-bigquery-sqlalchemy/issues/274.

The fix was to sync up with the version from sqlalchemy, to the extent possible given that it differs across different versions of SQLAlchemy.  

https://github.com/googleapis/python-bigquery-sqlalchemy/pull/275/commits/5e4074df9a0af98af4ecf470352f4ca9b3dcfd99#diff-efe1987a73ae5b54aa817dc8c70541025df89d9d08601c8ab33fa84ad8c09de7R266-R271 in particular.  This is needed so SQLAlchemy knows how to map subexpressions in queries (e.g. distinct) to columns.

The long term fix will be to stop copying this method.

File: sqlalchemy_bigquery/base.py [hunk 1]
```diff
@@ -258,9 +263,16 @@ def visit_column(
...
+            targets = (column, name, column.key) + result_map_targets
+            if getattr(column, "_tq_label", None):
+                # _tq_label was added in SQLAlchemy 1.4
+                targets += (column._tq_label,)
+
+            add_to_result_map(name, orig_name, targets, column.type)
 
         if is_literal:
+            # note we are not currently accommodating for
+            # literal_column(quoted_name('ident', True)) here
```
Comment (tseaver) [inline sqlalchemy_bigquery/base.py:275]: Is there an issue open for addressing that gap?

File: tests/system/test_sqlalchemy_bigquery.py [hunk 2]
```diff
@@ -691,3 +691,37 @@ def test_has_table(engine, engine_using_test_dataset, bigquery_dataset):
     assert engine_using_test_dataset.has_table(f"{bigquery_dataset}.sample") is True
 
     assert engine_using_test_dataset.has_table("sample_alt") is False
+
+
+def test_distinct_188(engine, bigquery_dataset, metadata):
```
Comment (tseaver) [inline tests/system/test_sqlalchemy_bigquery.py:696]: The `metadata` fixture doesn't look to be used here.

**Review by tseaver — COMMENTED** [review]
@jimfulton wrote:
> The long term fix will be to stop copying this method.

Is there an issue open to track the long-term fix?

**Comment by jimfulton** [comment]
> @jimfulton wrote:
> 
> > The long term fix will be to stop copying this method.
> 
> Is there an issue open to track the long-term fix?

https://github.com/googleapis/python-bigquery-sqlalchemy/issues/274

File: sqlalchemy_bigquery/base.py (hunk 1 above)
Comment (jimfulton) [reply sqlalchemy_bigquery/base.py:275]: That comment is part of the copy from sqlalchemy. I'll delete it. :/

File: tests/system/test_sqlalchemy_bigquery.py (hunk 2 above)
Comment (jimfulton) [reply tests/system/test_sqlalchemy_bigquery.py:696]: Right you are! I'll remove it.

File: sqlalchemy_bigquery/base.py (hunk 1 above)
Comment (jimfulton) [reply sqlalchemy_bigquery/base.py]: done

File: tests/system/test_sqlalchemy_bigquery.py (hunk 2 above)
Comment (jimfulton) [reply tests/system/test_sqlalchemy_bigquery.py]: done

**Comment by jimfulton** [comment]
@tseaver are you cool with this?

**Review by tseaver — APPROVED** [review]

**Pull request merged**